docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,126 @@
+    podAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchLabels:
+              foundationdb.org/distribution-key: "all-0"
+          topologyKey: fault_domain
+        weight: 1
+    # AntiAffinity to not schedule the Pod on the same fault domain where other Pods with a different distribution key are running
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+              key: "foundationdb.org/distribution-key"
+              operator: NotIn
+              values:
+              - "all-0"
+          topologyKey: fault_domain
+        weight: 1
+```
+
+The additional label and the `affinity` term enable the operator to spread Pods across fault domains.
+If `Required` is set to false this will be on a best effort base and can be useful if a fault domain is full but the Pod should be scheduled anyway.
+The following code snippet shows a possible implementation of the required structs:
+
+```golang
+type BinPack string
+
+const (
+    // BinPackAll defines that processes independent of their class should be bin packed
+    BinPackAll BinPack = "All"
+    // BinPackStateful defines that stateful processes should be bin packed and additionally stateless processes
+    BinPackStateful BinPack = "Stateful"
+    // BinPackClass only processes of the same class will be bin packed
+    BinPackClass BinPack = "Class"
+)
+
+// DistributionConfig
+type DistributionConfig struct {
+    // Enabled defines if the binpacking is enabled or not.
+    // Default: false
+    Enabled *bool
+    // DesiredFaultDomains desfines the number of desired fault domain.
+    // Should be greater than 0 is fault domain distribution is enabled.
+    // Default: 0
+    DesiredFaultDomains *int
+    // BinPack defines what processes should be bin packed to the fault domains.
+    // Default: "Class"
+    BinPack *BinPack
+    // Required defines if the affinity terms are required or preffered.
+    // Default: false
+    Required *bool
+}
+```
+
+The operator won't guarantee that the fault domains are equally used.
+Depending on the replacements for unreachable Pods we could use one fault domain more than another (or depending on the capacity),
+We also don't guarantee that we use exactly the number of `DesiredFaultDomains` e.g. when we spawn less processes than `DesiredFaultDomains` or when a fault domain is out of capacity and we are not using the `Required` with `true`.
+We also don't guarantee that the operator replaces any Pods that violate any constraint and we are not using the distribution with `Required` with `true` (`Required` with `true` would only ensure that we are not hurting the constraint).
+
+A change to `DesiredFaultDomains` will lead to a migration of all existing Pods in order to honor the new affinity term.
```

Comment:
How will we detect that this migration is necessary?

Reply:
By adding an additional label: https://github.com/FoundationDB/fdb-kubernetes-operator/pull/942/files#r721072454

Reply:
Could that lead to scenarios where we are migrating more than we have to?

Reply:
Yes, let me update that phrase and create an example.